### PR TITLE
[bugfix] fix flash-attention2 unavailable error for Ascend NPU

### PR DESCRIPTION
**Pull request by FightingZhen**
# What does this PR do?

#40002 cause Flash-Attention2 unavailable on Ascend NPU, **the core reason is package `flash-attn` unavailable on Ascend NPU, even can not be installed**, current implementation will cause **ImportError** on Ascend NPU.

This PR is committed for fixing this problem, remove redundant `apply_rotary_emb` function usage at the same time.

Fixes # (issue)
Not related.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by FightingZhen** [comment]
PR is ready for review @ArthurZucker @SunMarc @Rocketknight1 @ivarflakstad , cc @Byter-s @vasqu

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -77,16 +77,20 @@ def _lazy_imports(implementation: Optional[str]):
...
     is_fa3 = is_flash_attn_3_available()
-    if implementation == "flash_attention_2" or (implementation is None and is_fa2 and not is_fa3):
+
+    # Package `flash-attn` is unavailable on Ascend NPU, which will cause ImportError
+    # Flash-Attention2 related apis for Ascend NPU must be imported from `.integrations.npu_flash_attention` module
+    if implementation == "flash_attention_2" and is_torch_npu_available():
+        from .integrations.npu_flash_attention import npu_flash_attn_func as flash_attn_func
+        from .integrations.npu_flash_attention import npu_flash_attn_varlen_func as flash_attn_varlen_func
+        from .integrations.npu_flash_attention import pad_input, unpad_input
+    elif implementation == "flash_attention_2" or (implementation is None and is_fa2 and not is_fa3):
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py]: Am I understanding it right that the core reason it failed is because we had `implementation="flash_attention_2"` and it then tried to import from the fa package which is not available?

Wouldn't it be better to just change that conditional instead? E.g. 
- `if implementation == "flash_attention_2" or (implementation is None and is_fa2 and not is_fa3):` -> `if (implementation == "flash_attention_2" and is_fa2) or (implementation is None and is_fa2 and not is_fa3):`

Then all the rest doesn't need to be changed anymore?

File: src/transformers/modeling_flash_attention_utils.py [hunk 2]
```diff
@@ -77,16 +77,20 @@ def _lazy_imports(implementation: Optional[str]):
...
     is_fa2 = is_flash_attn_2_available()
     is_fa3 = is_flash_attn_3_available()
-    if implementation == "flash_attention_2" or (implementation is None and is_fa2 and not is_fa3):
+
+    # Package `flash-attn` is unavailable on Ascend NPU, which will cause ImportError
+    # Flash-Attention2 related apis for Ascend NPU must be imported from `.integrations.npu_flash_attention` module
+    if implementation == "flash_attention_2" and is_torch_npu_available():
+        from .integrations.npu_flash_attention import npu_flash_attn_func as flash_attn_func
+        from .integrations.npu_flash_attention import npu_flash_attn_varlen_func as flash_attn_varlen_func
+        from .integrations.npu_flash_attention import pad_input, unpad_input
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py]: Any reason to readd the copy pasted fa pad/unpad functions? `_pad` and `_unpad` are torch only variants which should just work the same but written in pure torch to avoid `einops` dependency; it was one of the main reasons this was removed in the previous PRs.

File: src/transformers/integrations/npu_flash_attention.py [hunk 3]
```diff
@@ -119,7 +233,7 @@ def npu_flash_attn_varlen_func(
     else:
         attn_mask_npu = get_attn_mask_npu(q.device)
         head_num = q.shape[1]
-        output = npu_fusion_attention(
+        output = torch_npu.npu_fusion_attention(
```
Comment (vasqu) [inline src/transformers/integrations/npu_flash_attention.py]: Just curious but is this needed? Meaning we need to import torch_npu and call with torch_npu.

Don't change it just interested in the import structure here :D

File: src/transformers/integrations/npu_flash_attention.py [hunk 4]
```diff
@@ -48,6 +51,117 @@ def is_npu_fa2_top_left_aligned_causal_mask():
...
+    """
+    # dim = hidden_states.shape[-1]
+    # output = torch.zeros((batch * seqlen), dim, device=hidden_states.device, dtype=hidden_states.dtype)
+    # output[indices] = hidden_states
+    output = index_put_first_axis(hidden_states, indices, batch * seqlen)
+    return rearrange(output, "(b s) ... -> b s ...", b=batch)
+
+
+# Copied from https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/bert_padding.py
+def unpad_input(hidden_states, attention_mask, unused_mask=None):
```
Comment (vasqu) [inline src/transformers/integrations/npu_flash_attention.py]: See my comment on the `_unpad` and `_pad` functions - would be really interested if they really didnt work and if they don't we should rather fix those methods imo (not necessarily this PR) or remove the einops dependencies with torch equivalent operations (e.g. reshape for rearragene etc).

**Review by vasqu — COMMENTED** [review]
Some initial comments, I'm not sure if we can hugely simplify this. That depends on if the _pad/_unpad functions really are not working as expected

File: src/transformers/integrations/npu_flash_attention.py (hunk 3 above)
Comment (FightingZhen) [reply src/transformers/integrations/npu_flash_attention.py]: Yeap, `torch_npu.npu_fusion_attention` api is the way to use FA2 on Ascend NPU.

File: src/transformers/modeling_flash_attention_utils.py (hunk 1 above)
Comment (FightingZhen) [reply src/transformers/modeling_flash_attention_utils.py]: `is_fa2` can not be used for Ascend NPU due to checking package `flash-attn` exists or not. Besides, for Ascend NPU, `implementation="flash_attention_2"` also eixsts. Therefore, `is_torch_npu_available` judgement can not be removed

Comment (vasqu) [reply src/transformers/modeling_flash_attention_utils.py]: But `is_fa2` checks for the availability of fa existing as a package in your env. Imo, the check just need to be changed to something like
```python 
    is_fa2 = is_flash_attn_2_available()
    is_fa3 = is_flash_attn_3_available()

    pad_input, unpad_input = _pad_input, _unpad_input
    if (implementation == "flash_attention_2" and is_fa2) or (implementation is None and is_fa2 and not is_fa3):
        from flash_attn import flash_attn_func, flash_attn_varlen_func
        from flash_attn.bert_padding import pad_input, unpad_input
    elif implementation == "flash_attention_2" and is_torch_npu_available():
        from .integrations.npu_flash_attention import npu_flash_attn_func as flash_attn_func
        from .integrations.npu_flash_attention import npu_flash_attn_varlen_func as flash_attn_varlen_func
    elif (implementation == "flash_attention_3" or implementation is None) and is_fa3:
        from flash_attn_interface import flash_attn_func, flash_attn_varlen_func
    # Kernels fallback
    else:
        flash_attn_func = getattr(implementation, "flash_attn_func", None)
        flash_attn_varlen_func = getattr(implementation, "flash_attn_varlen_func", None)
        if flash_attn_varlen_func is None or flash_attn_func is None:
            raise ValueError(
                f"Could not find the currently requested flash attention implementation at `{implementation}`."
                f"Make sure that you request a valid kernel from the hub, e.g. `kernels-community/flash-attn`."
            )

    return flash_attn_func, flash_attn_varlen_func, pad_input, unpad_input
```
Maybe I'm understanding something wrong, is the `is_fa_x` variable unreliable under npu?

**Comment by FightingZhen** [comment]
> Some initial comments, I'm not sure if we can hugely simplify this. That depends on if the _pad/_unpad functions really are not working as expected

@vasqu Thank you for your suggestions, I have tests `pad_input` and `unpad_input` funcs from your implementation, which is OK for Ascend NPU in terms of accuracy and performance, I will follow using your funcs. But the logical branch still can not be merged, detail reason can be found on above comments. Please help me have a check again, thanks :)

File: src/transformers/modeling_flash_attention_utils.py (hunk 1 above)
Comment (FightingZhen) [reply src/transformers/modeling_flash_attention_utils.py]: You are right, I misunderstood your suggestion, it truly works, thanks a lot! Code have been updated.

**Review by vasqu — APPROVED** [review]
LGTM and thank you! Also sorry that I broke things, flash attention really needs a lot of attention :sweat_smile: 

Letting Arthur take a final look but should be good to merge

File: src/transformers/models/qwen2_5_omni/modular_qwen2_5_omni.py
Comment (ArthurZucker) [inline src/transformers/models/qwen2_5_omni/modular_qwen2_5_omni.py:1]: good catch here!

**Review by ArthurZucker — APPROVED** [review]
Thanks, adding the for patch tag!
